### ISSUE TITLE
meson 0.56.0 errors out with Intel(R) oneAPI DPC++ Compiler 2021.1 (2020.10.0.1113)

### ISSUE THREAD

**Issue by ms178** [mesonbuild/meson#8113]
I am on Manjaro 20.2 and installed the new Intel(R) oneAPI DPC++ Compiler 2021.1 (2020.10.0.1113) which is now called "icx" for C and "icpx" for C++ (formerly known as: "icc" and "icpc").

I tried to compile some projects which use meson with makepkg from AUR and albeit having the compiler setup correctly, meson errors out with:

Using 'PKG_CONFIG_PATH' from environment with value: '/opt/intel/oneapi/mkl/latest/tools/pkgconfig'
Using 'PKG_CONFIG_PATH' from environment with value: '/opt/intel/oneapi/mkl/latest/tools/pkgconfig'
The Meson build system
Version: 0.56.0
Source dir: /home/marcus/Downloads/mesa-git/src/mesa
Build dir: /home/marcus/Downloads/mesa-git/src/_build
Build type: native build
Program python found: YES (/usr/bin/python)
Project name: mesa
Project version: 21.0.0-devel
Using 'CC' from environment with value: 'icx'

mesa/meson.build:21:0: ERROR: Unknown compiler(s): [['icx']]

I've used both methods described here (altered for the new 2021 edition) to setup the compiler with the same results as seen above: https://wiki.archlinux.org/index.php/Intel_C%2B%2B

I've also tried to modify the PKGBUILD file with adding CC, CXX and CC_LD options as described here: https://mesonbuild.com/howtox.html

The issue persists though. 

I've found a pull request which seems to be related to the underlying issue, but I am not an expert: https://github.com/mesonbuild/meson/pull/6827

**Comment by dcbaker** [comment]
This is probably due to changes in the output of icx compared to icc (the the C++ variants). I'll see if I can get icx installed this week and look at what's changed.

**Comment by ms178** [comment]
Thanks for taking a look. A lot has changed with the new release. They now use a modified Clang under the hood, icc/icpc lives on as the classic compiler (which this issue is not about, I haven't tried that newer release yet but icc/icpc did work with an older release some weeks ago). There is also another c++ compiler bundled with the new release, which is called "dpcpp" - which brings SYCL support.

**Comment by dcbaker** [comment]
I actually work for intel on the Linux graphics team (the oneAPI teams is my peers). so I actually no more about dpcpp than icx/icpx, :). I've thought a little bit about what sycl in meson would look like, but haven't gotten there yet. Like about a millino other things.

**Comment by ms178** [comment]
Re-visiting my issues, I've noticed https://github.com/mesonbuild/meson/pull/8764 - are there any news on supporting icx/icpx? Your comment seems to indicate that even though you are working for Intel, it doesn't seem to be much of a priority to support these new LLVM/Clang-based compilers in Meson soon...

**Comment by dcbaker** [comment]
This is going to sound bad, but that's not actually my job. I have access to the compilers as an employee, and i work on them as time allows.

**Comment by adibacco** [comment]
It is very strange that support for a new compiler from Intel is not already included. Is this support planned? Can someone elaborate on that?

**Comment by harry-van-haaren** [comment]
Hi All,

File `<meson>/compilers/detect.py` line `+512` replace
```
if 'clang' in out or 'Clang' in out:
```
with
```
if 'clang' in out or 'Clang' in out or 'oneAPI DPC++/C++ Compiler' in out:
```

Reasoning;
The OneAPI compiler gives a "clang compatible" binary, however the `--version` output is different. As the `--version` output is different, Meson doesn't detect/understand the compiler to be clang compatible one from a CLI/arguments etc point of view.

By adding the above string, we now detect the output of the OneAPI compiler to be similar to a "clang type" compiler, resulting in a successful configure/build.

Tested with [DPDK](http://dpdk.org/) on latest git, and with this one code-change in Meson, DPDK configures & builds correctly.

@Meson maintainers, please consider this change?

Regards, -Harry

**Comment by harry-van-haaren** [comment]
@ms178; your question above in https://github.com/mesonbuild/meson/issues/8113#issuecomment-883668524 explicitly mentions `icx` and `icpx` compiler binaries. The `--version` output of these binaries is the same as the above comment, so these are expected to also function correctly with the one-line change posted above.

**Comment by harry-van-haaren** [comment]
@dcbaker, @adibacco, @ms178, anybody available to provide a "tested-by"? It would be nice to get this merged.

**Comment by ms178** [comment]
@harry-van-haaren Thanks for your findings, but I am a bit reluctant to test out the current version of the oneAPI DPC++ Compiler as trying it out two weeks ago left me with an unbootable system. I will probably revisit this issue somewhen again in a quarter or two. But I hope Dylan Baker or some other project member can test/review your proposed changes in the meantime.

**Comment by harry-van-haaren** [comment]
@ms178 OK thanks for reporting back.

**Comment by vinsonlee** [comment]
@harry-van-haaren's suggestion in https://github.com/mesonbuild/meson/issues/8113#issuecomment-961945940 worked for me.


```
diff --git a/mesonbuild/compilers/detect.py b/mesonbuild/compilers/detect.py
index 03c5226d2..87f3bfd9f 100644
--- a/mesonbuild/compilers/detect.py
+++ b/mesonbuild/compilers/detect.py
@@ -501,7 +501,7 @@ def _detect_c_or_cpp_compiler(env: 'Environment', lang: str, for_machine: Machin
             return cls(
                 compiler, version, for_machine, is_cross, info, target,
                 exe_wrap, linker=linker)
-        if 'clang' in out or 'Clang' in out:
+        if 'clang' in out or 'Clang' in out or 'oneAPI DPC++/C++' in out:
             linker = None
 
             defines = _get_clang_compiler_defines(compiler)
```
`Tested-by: Vinson Lee <vlee@freedesktop.org>`

**Comment by harry-van-haaren** [comment]
Thanks for testing @vinsonlee!

**Comment by ms178** [comment]
@harry-van-haaren 

I've also tested it today with Xorg-Server-git, it works! Could you or Vinson Lee provide a MR to bring this issue to rest? ;)

`Tested-by: Marcus Seyfarth <m.seyfarth@gmail.com>`

**Comment by harry-van-haaren** [comment]
@ms178 thanks for testing! @vinsonlee, please go ahead and file a pull-request if you have time. If you'd like, a "Suggested-by:" with my name would be appreciated, but not essential.

**Comment by ms178** [comment]
@harry-van-haaren

My real name is mentioned in the last comment, I am happy either way.

**Comment by harry-van-haaren** [comment]
Doh - reading too fast here - apologies. Yes in that case it should certainly be included - thanks again.

**Comment by vinsonlee** [comment]
@harry-van-haaren @ms178 https://github.com/mesonbuild/meson/pull/9850

**Comment by harry-van-haaren** [comment]
Thanks @vinsonlee!

**Comment by ms178** [comment]
Intel just announced the ARC Pro cards on Siggraph, and sooner or later more people would like to use the bigger ARC cards (when they finally hit the market) together with the oneAPI toolchain. @harry-van-haaren @vinsonlee, as upstream devs would like to see some changes, what is the best strategy to move this over the finish line?

**Comment by harry-van-haaren** [comment]
Based on discussion in the pull-request (https://github.com/mesonbuild/meson/pull/9850), discussion is leaning towards a "full standalone compiler" concept inside Meson. A whole compiler impl is significantly more work, not something I can help with right now, unfortunately.

I see other projects are patching Meson to just include the above fix: https://github.com/spack/spack/pull/31155/commits/2d21bb3f1f34f218bd3023d741a058910d8d55a4

**Comment by ms178** [comment]
I understand, I'd prefer the more practical solution to merge the PR as is so that users can at least use that toolchain while the responsible team at Intel could perfect the details later on. As it stands now, users won't be able to use Meson for an unforseeable amount of time and given how slow things progressed since 2020, I have my doubts that we would see any fruits of that work anytime soon.

**Comment by alex-tee** [comment]
On Windows 10 and `Intel (R) oneAPI DPC++/C++ Compiler for applications running on Intel(R) 64, Version 2021.1 Beta Build 20200827`, after applying https://github.com/mesonbuild/meson/pull/9850 I get the following error (cannot detect the linker). Any ideas?

```
ERROR: Unable to detect linker for compiler `icx -Wl,--version`
```

![IMG_20221005_185900_1](https://user-images.githubusercontent.com/3407655/194196935-dae45e6c-f189-4395-8bb8-d4fe390900fc.jpg)

**Comment by alex-tee** [comment]
this MR from CMake may have some clues on how to get this into meson as well https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6806

**Comment by alex-tee** [comment]
For anyone having similar issues, using latest oneAPI on Windows (2022.2.0) you can use `CXX=clang-cl` and then meson will detect the compiler and linker. If it's not on the PATH you can add it manually: `C:\Program Files (x86)\Intel\oneAPI\compiler\2022.2.0\windows\bin-llvm`
According to README.txt in that directory it's not recommended to use the binaries from there directly and to use the bin directory instead (only compiler there is the `icx` binary which doesn't work with meson), but I haven't found any other solution yet.

**Comment by ms178** [comment]
As basic support landed recently and made it into the 0.64 release, I'll close this issue.
